### PR TITLE
[IMP] partner_invoicing_mode_at_shipping: Add grouping option per picking

### PR DESCRIPTION
**Pull request by rousseldenis**
Depends on :

- #1515

### CODE REVIEW

File: partner_invoicing_mode_at_shipping/README.rst [hunk 1]
```diff
@@ -33,6 +33,10 @@ It is based on `partner_invoicing_mode`.
 When this mode is selected the customer will be invoiced automatically on
 delivery of the goods.
 
+Another option is the 'One Invoice Per Picking'. That one is not compatible
```
Comment (jbaudoux) [inline partner_invoicing_mode_at_shipping/README.rst]: Shipping or Delivery. But not Picking
```suggestion
Another option is the 'One Invoice Per Delivery'. That one is not compatible
```

File: partner_invoicing_mode_at_shipping/models/res_partner.py [hunk 2]
```diff
@@ -11,3 +12,31 @@ class ResPartner(models.Model):
         selection_add=[("at_shipping", "At Shipping")],
         ondelete={"at_shipping": "set default"},
     )
+
+    one_invoice_per_picking = fields.Boolean(
```
Comment (jbaudoux) [inline partner_invoicing_mode_at_shipping/models/res_partner.py]:
```suggestion
    one_invoice_per_shipping = fields.Boolean(
```

File: partner_invoicing_mode_at_shipping/models/res_partner.py [hunk 3]
```diff
@@ -11,3 +12,31 @@ class ResPartner(models.Model):
...
+        help="Check this if you want to create one invoice per picking using the"
+        " partner invoicing mode that should be different than 'At Shipping'."
```
Comment (jbaudoux) [inline partner_invoicing_mode_at_shipping/models/res_partner.py]:
```suggestion
        help="Check this if you want to create one invoice per delivery. The invoice will be generated at shipping"
        " but will remain draft. The automatic validation can then be managed by the invoicing mode."
```

File: partner_invoicing_mode_at_shipping/models/res_partner.py [hunk 4]
```diff
@@ -11,3 +12,31 @@ class ResPartner(models.Model):
...
+    )
+    def _check_invoicing_mode_one_invoice_per_picking(self):
+        for partner in self:
+            if (
+                partner.invoicing_mode == "at_shipping"
+                and partner.one_invoice_per_picking
+            ):
+                raise ValidationError(
+                    _(
+                        "You cannot configure the partner %(partner)s with Invoicing Mode 'At Shipping' and 'One Invoice Per Picking'!"
```
Comment (jbaudoux) [inline partner_invoicing_mode_at_shipping/models/res_partner.py]:
```suggestion
                        "You cannot configure the partner %(partner)s with Invoicing Mode 'At Shipping' and 'One Invoice Per Shipping'!"
```

File: partner_invoicing_mode_at_shipping/models/res_partner.py [hunk 5]
```diff
@@ -11,3 +12,31 @@ class ResPartner(models.Model):
...
+                    ),
+                    partner=partner.name,
+                )
+            if partner.one_invoice_per_picking and partner.one_invoice_per_order:
+                raise ValidationError(
+                    _(
+                        "You cannot configure the partner %(partner)s with 'One Invoice Per Order' and 'One Invoice Per Picking'!"
```
Comment (jbaudoux) [inline partner_invoicing_mode_at_shipping/models/res_partner.py]:
```suggestion
                        "You cannot configure the partner %(partner)s with 'One Invoice Per Order' and 'One Invoice Per Shipping'!"
```

File: partner_invoicing_mode_at_shipping/models/res_partner.py [hunk 6]
```diff
@@ -11,3 +12,31 @@ class ResPartner(models.Model):
...
+
+    one_invoice_per_picking = fields.Boolean(
+        help="Check this if you want to create one invoice per picking using the"
+        " partner invoicing mode that should be different than 'At Shipping'."
+    )
+
+    @api.constrains(
+        "invoicing_mode", "one_invoice_per_picking", "one_invoice_per_order"
+    )
+    def _check_invoicing_mode_one_invoice_per_picking(self):
```
Comment (jbaudoux) [inline partner_invoicing_mode_at_shipping/models/res_partner.py]:
```suggestion
    def _check_invoicing_mode_one_invoice_per_shipping(self):
```

File: partner_invoicing_mode_at_shipping/models/res_partner.py [hunk 7]
```diff
@@ -11,3 +12,33 @@ class ResPartner(models.Model):
...
     )
+
+    one_invoice_per_picking = fields.Boolean(
+        help="Check this if you want to create one invoice per picking using the"
+        " partner invoicing mode that should be different than 'At Shipping'."
+    )
+
+    @api.constrains(
+        "invoicing_mode", "one_invoice_per_picking", "one_invoice_per_order"
+    )
```
Comment (jbaudoux) [inline partner_invoicing_mode_at_shipping/models/res_partner.py]: Why not making `one_invoice_per_shipping` computed store=True and set it False when invoicing_mode is "At shipping" or one_invoice_per_order is set ?

File: partner_invoicing_mode_at_shipping/models/res_partner.py [hunk 8]
```diff
@@ -11,3 +12,33 @@ class ResPartner(models.Model):
...
+    def _check_invoicing_mode_one_invoice_per_picking(self):
+        for partner in self:
+            if (
+                partner.invoicing_mode == "at_shipping"
+                and partner.one_invoice_per_picking
+            ):
+                raise ValidationError(
+                    _(
+                        "You cannot configure the partner %(partner)s with "
+                        "Invoicing Mode 'At Shipping' and 'One Invoice Per Picking'!",
```
Comment (jbaudoux) [inline partner_invoicing_mode_at_shipping/models/res_partner.py]:
```suggestion
                        "Invoicing Mode 'At Shipping' and 'One Invoice Per Shipping'!",
```

File: partner_invoicing_mode_at_shipping/models/res_partner.py [hunk 9]
```diff
@@ -11,3 +12,33 @@ class ResPartner(models.Model):
...
+                        partner=partner.name,
+                    ),
+                )
+            if partner.one_invoice_per_picking and partner.one_invoice_per_order:
+                raise ValidationError(
+                    _(
+                        "You cannot configure the partner %(partner)s with "
+                        "'One Invoice Per Order' and 'One Invoice Per Picking'!",
```
Comment (jbaudoux) [inline partner_invoicing_mode_at_shipping/models/res_partner.py]:
```suggestion
                        "'One Invoice Per Order' and 'One Invoice Per Shipping'!",
```

**Review by jbaudoux — CHANGES_REQUESTED** [review]

File: partner_invoicing_mode_at_shipping/models/res_partner.py (hunk 7 above)
Comment (rousseldenis) [reply partner_invoicing_mode_at_shipping/models/res_partner.py]: that could be an improvement yes

File: partner_invoicing_mode_at_shipping/models/sale_order.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        for order in self:
+            order.one_invoice_per_shipping = (
+                order.partner_invoice_id.one_invoice_per_shipping
+            )
+
+    def _get_generated_invoices(self, partition):
+        sales_to_validate_invoices = self.filtered(
+            lambda sale: sale.one_invoice_per_shipping
+            and any(invoice.state == "draft" for invoice in self.invoice_ids)
+        )
```
Comment (jbaudoux) [inline partner_invoicing_mode_at_shipping/models/sale_order.py]: @rousseldenis But if the invoice is already created and in draft, then the sale is not anymore in status "to invoice" and will be excluded from self with `_get_generate_invoices_state_domain`.

How can you enter this method? I think you won't even have a job for such sales

Comment (jbaudoux) [reply partner_invoicing_mode_at_shipping/models/sale_order.py]: The validation of those is already tackled by your `_validate_per_shipping_generated_invoices`, then I don't get why you need this hook ?

Comment (rousseldenis) [reply partner_invoicing_mode_at_shipping/models/sale_order.py]: I check this. It was part of the former implementation.

**Review by jbaudoux — APPROVED** [review]

**Comment by victorvermot** [comment]
@rousseldenis Hi I would love to use this pr could you please fix it ?

File: partner_invoicing_mode_at_shipping/hooks.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+                "one_invoice_per_shipping",
+                "sale.order",
+                "sale_order",
+                "boolean",
+                "boolean",
+                "partner_invoicing_mode_at_shipping",
+                False,
+            )
+        ]
+        openupgrade.add_fields(env, field_spec)
```
Comment (jbaudoux) [inline partner_invoicing_mode_at_shipping/hooks.py:21]: requires openupgradelib >= 3.5.1

**Comment by jbaudoux** [comment]
@rousseldenis dependency has been merged. Can you drop test-requirements.txt commit ?

**Review by lmignon — APPROVED** [review]

**Comment by jbaudoux** [comment]
@OCA/accounting-maintainers Could you merge this PR?

**Comment by rousseldenis** [comment]
@OCA/accounting-maintainers Gentle reminder

**Review by cyrilmanuel — APPROVED** [review]

**Comment by cyrilmanuel** [comment]
Hello @pedrobaeza, @sbidoul,  one of you can merge this pr ?

**Comment by pedrobaeza** [comment]
I don't use these modules, but merging it following current reviews:

/ocabot merge minor

**Pull request merged**